src/aead/aead.rs

```diff
@@ -76,37 +76,24 @@ impl OpeningKey {
     pub fn algorithm(&self) -> &'static Algorithm { self.key.algorithm() }
 }
 
-/// Authenticates and decrypts (&ldquo;opens&rdquo;) data in place.
-///
-/// The input is `in_out[in_prefix_len..]`; i.e. the input is the part of
-/// `in_out` after the prefix. When `open_in_place` returns `Ok(out_len)`, the
-/// decrypted output is `in_out[..out_len]`; i.e. the output has been written
-/// over the top of the prefix and the input. To put it a different way, the
-/// output overwrites the input, shifted by `in_prefix_len` bytes. To have the
-/// output overwrite the input without shifting, pass 0 as `in_prefix_len`.
-/// (The input/output buffer is expressed this way because Rust's type system
-/// does not allow us to have two slices, one mutable and one immutable, that
-/// reference overlapping memory at the same time.)
+/// Authenticates and decrypts (&ldquo;opens&rdquo;) data in place. When
+/// `open_in_place` returns `Ok(plaintext)`, the decrypted output is
+/// `plaintext`, which is a slice of `in_out`.
 ///
 /// C analog: `EVP_AEAD_CTX_open`
 ///
 /// Go analog: [`AEAD.Open`](https://golang.org/pkg/crypto/cipher/#AEAD)
-pub fn open_in_place(key: &OpeningKey, nonce: &[u8], in_prefix_len: usize,
-                     in_out: &mut [u8], ad: &[u8])
-                     -> Result<usize, error::Unspecified> {
+pub fn open_in_place<'a>(key: &OpeningKey, nonce: &[u8], ad: &[u8],
```

Comment:
> Is there a reason not to also break out the tag into its own argument here instead of requiring it be a suffix in in_out? Seems kinda asymmetric in its current state.

It is definitely asymmetric. But, I think that opening/sealing is conceptually asymmetric.

When sealing, you know how much plaintext you have, and you need to know that + the tag length to determine how much data to send.

When opening, you're almost always given the length of the length of `ciphertext + tag`, and you need to get the plaintext. You actually don't need to know how long the tag is, if you are given the plaintext length (or better, slice) as output.

Put another way, if we make `open_in_place()` symmetric with `seal_in_place()`, then it would actually require the caller to do more work, and that work would be error-prone. It would be more error-prone because the caller would need to split `ciphertext + tag`, being careful to handle the case where the input is shorter than the required tag length. Whereas, the way I've done it now seems less error-prone because this splitting is done internally.

OTOH, I do agree that the asymmetry is unfortunate. Maybe there's another solution I'm overlooking? Or maybe the benefit i described in this comment don't outweigh the pedagogical harm from having an asymmetric API? I do think some of the pedagogical harm of the asymmetry can be reduced with better documentation and/or better naming of the `in_out` parameter.